packages/backend/src/nest/registration/registration.service.ts

```diff
@@ -72,10 +70,76 @@ export class RegistrationService extends EventEmitter implements OnModuleInit {
     }
   }
 
-  public async registerUserCertificate(csr: string): Promise<void> {
-    const result = await issueCertificate(csr, this._permsData)
-    if (result?.cert) {
-      this.emit(RegistrationEvents.NEW_USER, { certificate: result.cert })
+  public async tryRegisterNextUserCertificates() {
+    if (!this.registrationEventInProgress) {
+      console.log("Registering next certificate")
+      this.registrationEventInProgress = true
+      const next = this.registrationEvents.shift()
+      if (next) {
+        await this.registerUserCertificates(next)
+      }
+      this.registrationEventInProgress = false
+
+      if (this.registrationEvents.length !== 0) {
+        setTimeout(this.tryRegisterNextUserCertificates.bind(this), 0)
+      }
     }
   }
+
+  // Apparently, JS will run each function to completion. So assuming
+  // we do not have multiple threads, this function should run to
+  // completion before it is called again. Because we take the CSRs
```

Comment:
this is inaccurate

JS will run the function synchronously until the first `await` statement. 
It's a cool excercise to look up the actual implementation of this in C-node. 
What happens underneath is that `await` is actually almost synonymous to `yield`. Async functions are really generators that "generate" Promises. Whenever Promise is returned, the handler applies to it's then() callback the logic that effectively calls `next(resultOfPreviousPromise)` on the generator. 
If you think about it in these terms it becomes obvious what will and what will not get interrupted

Reply:
Yea, this comment is outdated